File: go/internal/cmd/status.go
package cmd

import (
	"fmt"

	"github.com/spf13/cobra"

	"git.zib.de/cspiegel/threads/internal/git"
	"git.zib.de/cspiegel/threads/internal/thread"
	"git.zib.de/cspiegel/threads/internal/workspace"
)

var (
	statusCommit bool
	statusMsg    string
)

var statusCmd = &cobra.Command{
	Use:   "status <id> <new-status>",
	Short: "Change thread status",
	Args:  cobra.ExactArgs(2),
	RunE:  runStatus,
}

func init() {
	statusCmd.Flags().BoolVar(&statusCommit, "commit", false, "Commit after changing")
	statusCmd.Flags().StringVarP(&statusMsg, "m", "m", "", "Commit message")
}

func runStatus(cmd *cobra.Command, args []string) error {
	ws := getWorkspace()
	ref := args[0]
	newStatus := args[1]

	file, err := workspace.FindByRef(ws, ref)
	if err != nil {
		return err
	}

	t, err := thread.Parse(file)
	if err != nil {
		return err
	}

	oldStatus := t.Status()

	if err := t.SetFrontmatterField("status", newStatus); err != nil {
		return err
	}

	if err := t.Write(); err != nil {
		return err
	}

	fmt.Printf("Status changed: %s → %s (%s)\n", oldStatus, newStatus, file)

	if statusCommit {
		msg := statusMsg
		if msg == "" {
			msg = git.GenerateCommitMessage(ws, []string{file})
		}
		if err := git.AutoCommit(ws, file, msg); err != nil {
			return err
		}
	} else {
		fmt.Printf("Note: Thread %s has uncommitted changes. Use 'threads commit %s' when ready.\n", ref, ref)
	}

	return nil
}

File: rust/src/cmd/status.rs
use std::path::Path;

use clap::Args;

use crate::git;
use crate::thread::{self, Thread};
use crate::workspace;

#[derive(Args)]
pub struct StatusArgs {
    /// Thread ID or name reference
    id: String,

    /// New status
    new_status: String,

    /// Commit after changing
    #[arg(long)]
    commit: bool,

    /// Commit message
    #[arg(short = 'm', long)]
    m: Option<String>,
}

pub fn run(args: StatusArgs, ws: &Path) -> Result<(), String> {
    if !thread::is_valid_status(&args.new_status) {
        return Err(format!(
            "Invalid status '{}'. Must be one of: idea, planning, active, blocked, paused, resolved, superseded, deferred, reject",
            args.new_status
        ));
    }

    let file = workspace::find_by_ref(ws, &args.id)?;

    let mut t = Thread::parse(&file)?;
    let old_status = t.status().to_string();

    t.set_frontmatter_field("status", &args.new_status)?;
    t.write()?;

    println!("Status changed: {} → {} ({})", old_status, args.new_status, file.display());

    if args.commit {
        let msg = args.m.unwrap_or_else(|| {
            git::generate_commit_message(ws, &[file.to_string_lossy().to_string()])
        });
        git::auto_commit(ws, &file, &msg)?;
    } else {
        println!(
            "Note: Thread {} has uncommitted changes. Use 'threads commit {}' when ready.",
            args.id, args.id
        );
    }

    Ok(())
}
